### DISCUSSION TITLE
RTK Query Sustainable Setup As Codebase Grows In Complexity

### DISCUSSION THREAD

**Discussion by agusterodin** [reduxjs/redux-toolkit#1295] [Q&A] [answered]
I am loving RTK query for how powerful it is. My code, however, is getting really heavy. I feel as if many people would get turned off to RTK query solely for how complicated and dense the code ends up looking. Unless I am missing something, it seems like the convenience of TypeScript comes at the cost of everything needing to be defined in single object.  

It doesn't help that RTK Query feels like a fully fledged library itself and deserves its own separate documentation site. I am sure that the maintainers put A LOT of thought into this decision and have good reasons, especially since RTK query is internally built off of other parts of RTK. Regardless, the docs feel cluttered now.

I have no problems understanding what is going on in both the docs and my code but I could imagine people who are newer to JS/TS/React/Redux being completely overwhelmed. I mean this in the most constructive possible way. This is my gut reaction to everything I have experienced so far. I never had this overwhelmed feeling when using other parts of RTK (like createSlice).

I have been experimenting with splitting the endpoints up into separate files like this. The plan is to have one file per corresponding controller in the backend. This is my initial attempt:

api/index.ts

```import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { EndpointBuilder } from '@reduxjs/toolkit/dist/query/endpointDefinitions'
import campaignEndpoints from './campaign'

const reducerPath = 'api' as const

const tagTypes = ['campaignMetrics'] as const

const baseQuery = fetchBaseQuery({
  baseUrl: '/',
  prepareHeaders: headers => {
    const accessToken = localStorage.getItem('access_token')
    if (accessToken) headers.set('authorization', `Bearer ${accessToken}`)
    return headers
  }
})

export const campaignsApi = createApi({
  reducerPath,
  baseQuery,
  tagTypes,
  endpoints: builder => ({
    ...campaignEndpoints(builder)
  }),
  refetchOnFocus: true,
  refetchOnReconnect: false
})

export type CampaignsApiEndpointBuilder = EndpointBuilder<typeof baseQuery, typeof tagTypes[number], typeof reducerPath>
```

api/campaign.ts

```
import { campaignsApi, CampaignsApiEndpointBuilder } from './index'

export default (builder: CampaignsApiEndpointBuilder) => {
  return {
    getCampaignMetricsById: builder.query<Metrics, number>({
      query: id => `campaign/${id}/metrics`,
      providesTags: ['campaignMetrics']
    }),
    getCampaignMetadataById: builder.query<CampaignMetadata, number>({
      query: id => `campaign/${id}`,
      providesTags: ['campaignMetrics']
    }),
    setCampaignActive: builder.mutation<void, { id: number; active: boolean }>({
      query: ({ id, active }) => ({
        url: `campaign/activate/${id}`,
        method: RequestMethod.Put,
        body: { active }
      }),
      invalidatesTags: ['campaignMetrics'],
      async onQueryStarted({ id, active }, { dispatch, queryFulfilled }) {
        const patchResult = dispatch(
          campaignsApi.util.updateQueryData('getCampaignMetadataById', id, draft => {
            Object.assign(draft, { active })
          })
        )
        try {
          await queryFulfilled
        }
        catch {
          patchResult.undo()
        }
      }
    })
  }
}

export const { useGetCampaignMetadataByIdQuery, useGetCampaignMetricsByIdQuery, useSetCampaignActiveMutation } = campaignsApi
```

TS inference stuff works perfectly. Unfortunately cyclical dependencies bite me in the ass and the app crashes when I run it. The console says Object(...) is not a function since campaignEndpoints is not defined when createApi gets executed.

I'm sure there are ways to restructure this to avoid circular dependencies. Does anyone have suggestions on a nice structure?

Also if there is an elegant way to structure things to accommodate a growing codebase it may be worth adding as an example in the docs.

**Comment by markerikson** [comment, marked as answer]
My first quick thought is that you should be using the patterns shown in the "Code Splitting" page to split up the endpoint definitions, even if you're not actually trying to literally code-split:

https://redux-toolkit.js.org/rtk-query/usage/code-splitting

**Reply by agusterodin** [reply]
Oh wow didn't see that. I feel like a dumbass.

**Reply by agusterodin** [reply]
I'm the kind of dumbass that complains about something when there is a page in the docs that addresses it.

The name 'code splitting' is what made me miss it.

**Reply by markerikson** [reply]
Hah, no worries :)  It's not immediately obvious that this applies to generally breaking up logic across multiple files.

Agreed with your initial points that A) there's a lot of material, B) there can be a lot of code, C) it can be overwhelming, and D) RTKQ is kinda like its own lib.  But yeah, we did have good reasons for all the organizational decisions we've made so far:

- Having RTKQ embedded in the RTK package means that it's easier for people to adopt, especially in an enterprise-type environment. There's no "separate package" to install, just "update to the latest RTK"
- We debated the docs organization quite a bit before settling on this setup. We were also looking at having "Usage >RTKQ" and "API > RTKQ", but decided it was worth highlighting RTKQ as its own top-level thing.

The code can definitely add up here, yeah, but on the other hand... just think of how many thunks and reducers this is replacing :)  (and for that matter, this could all potentially be auto-generated based on an OpenAPI or GraphQL schema)

**Reply by zzzbra** [reply]
In OP's defense I wasn't able to find this section after some thirty, forty minutes of searching as well.

**Reply by markerikson** [reply]
Any suggestions on other ways to organize this info?

**Reply by agusterodin** [reply]
Fortunately @phryneas may have figured out a solution that doesn't (to my understanding) require use of injectEndpoints anymore https://github.com/reduxjs/redux-toolkit/pull/1686 !

Not that there's anything wrong with injectEndpoints. It works very well.

Edit: not sure if this applies to RTK query or just other parts of RTK.

**Reply by zzzbra** [reply]
@markerikson I can see now that it's mentioned [here](https://redux.js.org/tutorials/essentials/part-8-rtk-query-advanced#injecting-endpoints) in the Advanced RTK Query section of the Redux essentials doc. I'm not sure why I missed it, I think "code-splitting" is a bit overloaded and that I didn't realize it applied here. 

Something else that I have not seen (yet, probably) is a mention of the relation between `services` and `features` in the ideal organization of a large app leveraging RTKQ. Am I just missing that? Once you begin to break out the api slices into separate files, I'm unclear if I should continue to group them one way or the other.

**Reply by markerikson** [reply]
We don't have much specific guidance on file structure beyond "use feature folders, and put logic for a given feature into a single slice file":

https://redux.js.org/style-guide/style-guide#structure-files-as-feature-folders-with-single-file-logic

We also specifically recommend having a single API slice _definition_ per app unless you truly have multiple completely different servers the app is talking to.  However, the _implementation_ of that API slice can be split across multiple files via `injectEndpoints`.

As to where you put those split implementations, that's totally up to you.

**Reply by agusterodin** [reply]
My personal rule of thumb is each controller on our backend has a corresponding api slice (each in it's own dedicated file). We put these in a folder called "api" in the root of our project. Here is an example how I structured an API for a personal project of mine:

[api.zip](https://github.com/reduxjs/redux-toolkit/files/7479381/api.zip)

All other Redux stuff (slices, selectors, store initialization, etc) goes in a folder called "state" in the root of my project.

This is highly subjective, so find what works best for your project :D

**Reply by rozendale** [reply]
As a beginner in REACT, REDUX, TS and even JS, it is indeed overwhelming. However, I start to get along and RTK query seems to me amazing, hopefully it won't change.

**Reply by zzzbra** [reply]
One point that it might be helpful to emphasize in the docs, although it is said already in so many words, is that using RTKQ will *not* result in the data it cache's being stored in the same type of global state tree that older users of Redux probably think of first when they think of the library. Of course, one can put it there manually but then that would undermine the entire rationale behind creating RTKQ in the first place.

**Reply by Jonesus** [reply]
> Any suggestions on other ways to organize this info?

I had similar difficulties finding the "Code Splitting" part of the documentation initially as well, as "code splitting" in my head maps more to bundle size optimization (which is exactly what you start the docs page with), but almost equally as important IMO is th easpect of organizing a project's codebase. Maybe the title could be changed to something like "Code Splitting & Organization", "Code Splitting & Project Structure", or have another page that highlights the `injectEndpoints`'s capability of being used as a codebase organization feature in addition to bundle splitting?

**Comment by Chococrok** [comment]
Hello, I really like @agusterodin structure. The problem is that the utilities are only available on the api slice and not on the builder.

An other thing that is annoying is that injectEndpoints does not modify the api slice in place so I end up doing something like:
```ts

const withSomeEndpoints = addSomeEndpoints(baseApi);
const withOtherEndpoints = addOtherEndpoints(withSomeEndpoints);
...
export const fullApi = addAgainOtherEndpoints(withOtherEndpoints );
export {...} = fullApi;
...
```

But maybe I missed something ?

**Reply by phryneas** [reply]
It modifes everything in place, there is just no way the types can reflect that, especially not across file boundaries. There is no real value in building a `fullApi` at any place. Just inject into the empty api in all your files and export what you need from the file that injected.
